<a href="https://www.kaggle.com/code/abbas829/pandas-data-story?scriptVersionId=294665172" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

<a href="https://www.kaggle.com/code/abbas829/pandas-01b?scriptVersionId=292637312" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# 🔍 The Data Detective: Mastering Exploratory Data Analysis with Pandas

## Uncovering Hidden Insights in User Behavior using Python

**Mission Briefing:** In the fast-paced world of tech startups, data is the ultimate evidence. You've just been hired as the **Lead Data Investigator**. Your first assignment? A mysterious dataset containing raw user profiles. Your boss needs to know *who* these users are before the marketing team can launch their next big campaign.

As the **Data Detective**, you will use **Pandas**—the most powerful library in the Python ecosystem—to forensicially examine this data, identify patterns, and solve the mystery of your user base.

---

## 🛠️ Step 1: Gathering the Forensic Toolkit

Every investigator needs a reliable set of tools. For Python-based data forensics, we rely on the industry standards: `Pandas` for manipulation and `NumPy` for numerical operations.

> [!TIP]
> Standard convention is to import pandas as `pd` and numpy as `np`. This keeps our code clean and matches universal documentation styles.

In [1]:
import numpy as np
import pandas as pd

## 📍 Step 2: Locating the Evidence (Data Acquisition)

Our intelligence report points to a remote server where the raw logs are stored. We need to define the source URL before we can begin the extraction.

*Action: Define the source URI for the user dataset.*

In [2]:
url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user'

## 📁 Step 3: Opening the Case File

Now, we ingest the data into a Pandas **DataFrame**. We've noticed the file uses a pipe `|` delimiter rather than the standard comma, and we want to use the `user_id` as our primary key (index).

```mermaid
graph LR
A[Raw Data Link] --> B{Pandas Read}
B --> C[Structured DataFrame]
C --> D[Data Detective Dashboard]
```

> [!IMPORTANT]
> Using `index_col` during the loading phase saves us from having to reset the index later, keeping our memory usage efficient.

In [3]:
users = pd.read_csv(url, sep='|', index_col='user_id')
users.head()

age gender  occupation zip_code
user_id                                 
1         24      M  technician    85711
2         53      F       other    94043
3         23      M      writer    32067
4         24      M  technician    43537
5         33      F       other    15213

## 👀 Step 4: First Glance at the Suspects

A detective never dives in without a quick overview. We'll examine the first 25 entries to identify visible patterns or potential anomalies in the data entry.

| Feature | Description |
| :--- | :--- |
| `age` | User's age in years |
| `gender` | M/F designation |
| `occupation` | Self-reported job title |
| `zip_code` | Geographic location marker |

In [4]:
users.head(25)

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

## 🔍 Step 5: Checking the Trail's End

Does the data remain consistent until the very last row? We check the final 10 entries to verify there are no trailing errors or mismatched formats.

*Action: Inspect the tail of the dataset.*

In [5]:
users.tail(10)

age gender     occupation zip_code
user_id                                    
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

## 📊 Step 6: Measuring the Population Scale

How large is our investigation? Knowing the total number of observations is critical for statistical significance.

**Metric:** Total Number of Users

In [6]:
print(f'Total Users Investigated: {len(users)}')

Total Users Investigated: 943


## 🏷️ Step 7: Inventorying User Features

What specific data points do we have on each suspect? Counting the columns allows us to plan our multidimensional analysis.

*Action: Count and list the feature set.*

In [7]:
print(f'Number of Features: {len(users.columns)}')
print(f'Feature Names: {list(users.columns)}')

Number of Features: 4
Feature Names: ['age', 'gender', 'occupation', 'zip_code']


## 🧬 Step 8: Analyzing Data DNA (Types)

Understanding data types (`dtypes`) is like checking the DNA of your evidence. Are we looking at integers, text (objects), or categories?

> [!NOTE]
> In Pandas, text columns are often listed as `object`. For modern analysis, we could convert these to `category` to save memory.

In [8]:
users.dtypes

age            int64
gender        object
occupation    object
zip_code      object
dtype: object

## 👮 Step 9: Deep Dive into Occupations

The boss is curious about the professional background of our users. Let's isolate the `occupation` column for specialized analysis.

*Action: Extract and view the occupation profile.*

In [9]:
users.occupation

user_id
1         technician
2              other
3             writer
4         technician
5              other
           ...      
939          student
940    administrator
941          student
942        librarian
943          student
Name: occupation, Length: 943, dtype: object

## 🎓 Step 10: Identifying Diversity (Unique Jobs)

Is our user base concentrated in one field, or spread across various industries? Counting the unique entries tells us the level of diversity.

*Action: Count distinct occupations.*

In [10]:
unique_jobs = users['occupation'].nunique()
print(f'Distinct Careers Identified: {unique_jobs}')

Distinct Careers Identified: 21


## 🏆 Step 11: The Most Frequent Actor

Which occupation dominates the ecosystem? Identifying the mode (most common value) helps tailer marketing strategies.

*Action: Find the top occupation.*

In [11]:
most_common = users['occupation'].value_counts().idxmax()
print(f'Primary User Segment: {most_common}')

Primary User Segment: student


## 📉 Step 12: The Statistical Summary

Let's run a full statistical profile on the numeric data. This gives us the mean, standard deviation, and quartiles of our user ages.

*Action: Generate descriptive statistics.*

In [12]:
users.describe()

age
count  943.000000
mean    34.051962
std     12.192740
min      7.000000
25%     25.000000
50%     31.000000
75%     43.000000
max     73.000000

## 📋 Step 13: The Holistic Investigation Report

By including all columns in our summary, we can even see metrics for text columns, such as 'unique' and 'top' frequency.

*Action: View the full summary for all features.*

In [13]:
users.describe(include='all')

age gender occupation zip_code
count   943.000000    943        943      943
unique         NaN      2         21      795
top            NaN      M    student    55414
freq           NaN    670        196        9
mean     34.051962    NaN        NaN      NaN
std      12.192740    NaN        NaN      NaN
min       7.000000    NaN        NaN      NaN
25%      25.000000    NaN        NaN      NaN
50%      31.000000    NaN        NaN      NaN
75%      43.000000    NaN        NaN      NaN
max      73.000000    NaN        NaN      NaN

## ⏳ Step 14: Analyzing the Demographics (Average Age)

Understanding the age distribution is vital for user experience design. 

*Action: Calculate the mean age.*

In [14]:
mean_age = users['age'].mean()
print(f'Average User Age: {mean_age:.1f} years')

Average User Age: 34.1 years


## 🔍 Step 15: Identifying Outliers

Which ages are the rarest in our population? Finding the least common entries helps identify niche groups.

*Action: Examine the frequency of rare age segments.*

In [15]:
users['age'].value_counts().tail(5)

age
7     1
66    1
11    1
10    1
73    1
Name: count, dtype: int64

--- 

# 🏁 Case Closed: Final Intelligence Briefing

**Mission Accomplished.** You've successfully navigated the complexities of raw data and transformed it into actionable intelligence.

### Key Findings Summary:
- 👥 **Population Size:** 943 Users
- 💼 **Job Diversity:** 21 Distinct Occupations
- 🎓 **Dominant Group:** Students
- 🎂 **Average Age:** ~34.1 years

This initial exploration forms the bedrock of any successful Machine Learning project. Without knowing your data, you cannot model it. 

**Well done, Detective.**

---

## 👤 Lead Investigator Details

**Author:** Tassawar Abbas  
**Email:** abbas829@gmail.com  